core/templates/hash_map.h

```diff
@@ -93,180 +94,189 @@ class HashMap {
-		const uint32_t capacity = hash_table_size_primes[capacity_index];
-		const uint64_t capacity_inv = hash_table_size_primes_inv[capacity_index];
+	void _insert_with_hash_and_element(uint32_t p_hash, HashMapElement<TKey, TValue> *p_value) {
+		const uint32_t local_capacity = capacity;
+		uint32_t pos = p_hash & local_capacity;
+
+		if (hashes[pos] == EMPTY_HASH) {
+			elements[pos] = p_value;
+			hashes[pos] = p_hash;
+			num_elements++;
+			return;
+		}
+
 		uint32_t hash = p_hash;
 		HashMapElement<TKey, TValue> *value = p_value;
-		uint32_t distance = 0;
-		uint32_t pos = fastmod(hash, capacity_inv, capacity);
-
+		uint32_t distance = 1;
+		pos = (pos + 1) & local_capacity;
 		while (true) {
 			if (hashes[pos] == EMPTY_HASH) {
 				elements[pos] = value;
 				hashes[pos] = hash;
-
 				num_elements++;
-
+#ifdef DEV_ENABLED
+				if (unlikely(distance > 12)) {
+					WARN_PRINT("Excessive collision count (" +
+							itos(distance) + "), is the right hash function being used?\nClass Name:" + __PRETTY_FUNCTION__);
+				}
+#endif
 				return;
 			}
 
 			// Not an empty slot, let's check the probing length of the existing one.
-			uint32_t existing_probe_len = _get_probe_length(pos, hashes[pos], capacity, capacity_inv);
+			uint32_t existing_probe_len = _get_probe_length(pos, hashes[pos], local_capacity);
 			if (existing_probe_len < distance) {
 				SWAP(hash, hashes[pos]);
 				SWAP(value, elements[pos]);
 				distance = existing_probe_len;
 			}
 
-			pos = fastmod((pos + 1), capacity_inv, capacity);
+			pos = (pos + 1) & local_capacity;
 			distance++;
 		}
 	}
 
-	void _resize_and_rehash(uint32_t p_new_capacity_index) {
-		uint32_t old_capacity = hash_table_size_primes[capacity_index];
-
-		// Capacity can't be 0.
-		capacity_index = MAX((uint32_t)MIN_CAPACITY_INDEX, p_new_capacity_index);
+	void _resize_and_rehash(uint32_t p_new_capacity) {
+		const uint32_t real_old_capacity = capacity + 1;
 
-		uint32_t capacity = hash_table_size_primes[capacity_index];
+		// Capacity can't be 0 and must be 2^n - 1.
+		capacity = MAX(4u, p_new_capacity);
+		const uint32_t real_capacity = next_power_of_2(capacity);
+		capacity = real_capacity - 1;
 
 		HashMapElement<TKey, TValue> **old_elements = elements;
 		uint32_t *old_hashes = hashes;
 
-		num_elements = 0;
-		hashes = reinterpret_cast<uint32_t *>(Memory::alloc_static(sizeof(uint32_t) * capacity));
-		elements = reinterpret_cast<HashMapElement<TKey, TValue> **>(Memory::alloc_static(sizeof(HashMapElement<TKey, TValue> *) * capacity));
+		hashes = reinterpret_cast<uint32_t *>(Memory::alloc_static(sizeof(uint32_t) * real_capacity));
+		elements = reinterpret_cast<HashMapElement<TKey, TValue> **>(Memory::alloc_static(sizeof(HashMapElement<TKey, TValue> *) * real_capacity));
 
-		for (uint32_t i = 0; i < capacity; i++) {
-			hashes[i] = 0;
-			elements[i] = nullptr;
-		}
+		memset(hashes, EMPTY_HASH, real_capacity * sizeof(uint32_t));
```

Comment:
Since `EMPTY_HASH` is `uint32_t`, this line is not formally correct. It's taking advantage of the fact that all the bytes in the value are the same (0, in this case). I'm not against using `memset()` it it proves to be more efficient than a plain, old loop (i.e., if the compiler can't realize it can use SIMD). However, I'd suggest documenting that this line (and the other `memset()` one) is making the assumption, or ensuring future maintainers will realize they broke the assumption somehow.

For instance by adding something like this below declaring `EMPTY_HASH`:
`static_assert(EMPTY_HASH <= 255, "Value must fit in a byte for the memcpy() optimization.");`

Reply:
It seems that EMPTY_HASH should always be 0. Because `memset` sets the number by byte. For example this code
```c++
uint32_t i = 0;
memset(&i, 1, 4);
```
will return 16843009.

Reply:
So I will make such a check
```
static_assert(EMPTY_HASH == 0, "EMPTY_HASH must always be 0 for the memcpy() optimization.");
```

Reply:
Yeah, 0 is the simplest.

Just for completeness, what I was considering (without actually explaining my point very well) is that, theoretically, you could use any byte value, but you'd need two constants: one for the byte value and one for the composite 32-bit result (`(EMPTY_HASH_BYTE << 24) | (EMPTY_HASH_BYTE << 16) | (EMPTY_HASH_BYTE << 8) | EMPTY_HASH_BYTE`). And that's actually terrible, so yes, zero is the winner here!